app/grandchallenge/algorithms/models.py

```diff
@@ -629,6 +649,118 @@ def create(
+        Checks if there are existing CIVs for the provided data and returns those.
+
+        Parameters
+        ----------
+        civ_data
+            A dictionary with interface slugs as keys and CIV values as values.
+
+        Returns
+        -------
+        A list of ComponentInterfaceValues
+
+        """
+        existing_civs = []
+        for interface, value in civ_data.items():
+            if isinstance(value, ComponentInterfaceValue):
+                existing_civs.append(value)
+            elif isinstance(value, Image):
+                ci = ComponentInterface.objects.get(slug=interface)
+                try:
+                    civ = ComponentInterfaceValue.objects.filter(
+                        interface=ci, image=value
+                    ).get()
+                    existing_civs.append(civ)
+                except ObjectDoesNotExist:
+                    continue
+            elif isinstance(
+                value, (RawImageUploadSession, UserUpload, QuerySet)
+            ):
+                # uploads will create new CIVs, so ignore these
+                continue
+            else:
+                # values can be of different types
+                ci = ComponentInterface.objects.get(slug=interface)
+                try:
+                    civ = ComponentInterfaceValue.objects.filter(
+                        interface=ci, value=value
+                    ).get()
+                    existing_civs.append(civ)
+                except ObjectDoesNotExist:
+                    continue
+
+        return existing_civs
+
+    def get_jobs_with_same_inputs(
+        self, *, data, algorithm_image, algorithm_model
+    ):
+        civ_data = {
+            k: v for k, v in data.items() if k not in self.non_interface_fields
+        }
+        existing_civs = self.retrieve_existing_civs(civ_data=civ_data)
+        unique_kwargs = {
+            "algorithm_image": algorithm_image,
+        }
+        input_interface_count = algorithm_image.algorithm.inputs.count()
+        if algorithm_model:
+            unique_kwargs["algorithm_model"] = algorithm_model
+        else:
+            unique_kwargs["algorithm_model__isnull"] = True
+        existing_jobs = (
+            Job.objects.filter(**unique_kwargs)
+            .annotate(
+                inputs_match_count=Count(
+                    "inputs", filter=Q(inputs__in=existing_civs)
+                )
+            )
+            .filter(inputs_match_count=input_interface_count)
+        )
+        return existing_jobs
+
+    def create_job_and_process_inputs(
+        self,
+        *,
+        data,
+        extra_viewer_groups=None,
+        extra_logs_viewer_groups=None,
+    ):
+        civ_data = {
+            k: v for k, v in data.items() if k not in self.non_interface_fields
+        }
+        non_civ_data = {
+            k: v for k, v in data.items() if k in self.non_interface_fields
+        }
+        job = Job.objects.create(
+            **non_civ_data,
```

Comment:
There could be a bug here. What if someone introduces a CI with the slug the same as one of the private field for the job? It is safer to be explicit about what is being accepted by doing a direct lookup of the fields that are accepted here.

Reply:
Ok, good point. I will change this to use direct lookups for the non civ fields. And in order to tell the interface value from the private field value in the case you describe, the callers of this method here should provide `civ_data` and `non_civ_data` rather than splitting it here.

Reply:
To prevent the issue you describe above, we actually need to prepend the interface fields in the form with something to clearly mark them as separate from the object fields. I will add an underscore.